File: typescript/day02/main.ts
import fs from 'fs'

const CUBES_MAX: Map<string, number> = new Map([
    ["red", 12],
    ["green", 13],
    ["blue", 14]
])
function load_game_data(filename: string): Map<number, Map<string, number>> {
    let game_data = new Map<number, Map<string, number>>();
    let data = fs.readFileSync(filename)
    data.toString().split('\n').forEach((line: string) => {
        let [game, draws] = line.split(':')
        let [_, game_index_str] = game.split(' ')
        let game_index = parseInt(game_index_str)
        let cubes_max: Map<string, number> = new Map([
            ["red", 0],
            ["green", 0],
            ["blue", 0]
        ])
        for (let draw of draws.split(';')) {
            for (let cubes of draw.split(',')) {
                let [cube_count_str, cube_color] = cubes.trim().split(' ')
                let cube_count = parseInt(cube_count_str)
                cubes_max.set(cube_color, Math.max(cubes_max.get(cube_color) ?? 0, cube_count))
            }
        }
        game_data.set(game_index, cubes_max)
    })
    return game_data
}

function solution_part_1() {
    let possible_games_sum = 0
    for (let [game_index, game] of load_game_data("input.txt")) {
        if (
          game.get("red")! <= CUBES_MAX.get("red")! &&
            game.get("green")! <= CUBES_MAX.get("green")! &&
            game.get("blue")! <= CUBES_MAX.get("blue")!
        ) {
            possible_games_sum += game_index
        }
    }
    console.log(possible_games_sum)
}

function solution_part_2() {
    let power_sum = 0
    for (let [_, game] of load_game_data("input.txt")) {
        power_sum += game.get("red")! * game.get("green")! * game.get("blue")!
    }
    console.log(power_sum)
}

solution_part_1()
solution_part_2()

File: Rust/day02/src/main.rs
use std::collections::HashMap;
use std::fs;
use lazy_static::lazy_static;

fn parse_game_data(data: &String) -> HashMap<i32, Vec<HashMap<String, i32>>> {
    let mut games : HashMap<i32, Vec<HashMap<String, i32>>> = HashMap::new();
    for line in data.lines() {
        // split line
        let split = line.split(":").collect::<Vec<&str>>();
        let game_index = split[0].split(" ").collect::<Vec<&str>>()[1].parse::<i32>().unwrap();
        let game_data_split = split[1].split(";").collect::<Vec<&str>>();
        let mut game = Vec::new();
        for single_draw in game_data_split {
            let mut draw = HashMap::new();
            let single_draw_split = single_draw.split(",").collect::<Vec<&str>>();
            for color in single_draw_split {
                let color_split = color.trim().split(" ").collect::<Vec<&str>>();
                let color_name = String::from(color_split[1]);
                let color_count = color_split[0].parse::<i32>().unwrap();
                draw.insert(color_name, color_count);
            }
            game.push(draw);
        }
        games.insert(game_index, game);
    }
    return games
}

lazy_static! {
    static ref GIVEN_CUBES: HashMap<String, i32> = {
        let mut m = HashMap::new();
        m.insert("red".to_string(), 12);
        m.insert("green".to_string(), 13);
        m.insert("blue".to_string(), 14);
        m
    };
}

fn test_game(game: &Vec<HashMap<String, i32>>) -> bool {
    for draw in game {
        for (color, count) in draw {
            if GIVEN_CUBES.contains_key(color) {
                let current_count = GIVEN_CUBES.get(color).unwrap();
                if current_count < count {
                    return false
                }
            } else {
                return false
            }
        }
    }
    return true
}
fn solution_part_1() {
    let data = fs::read_to_string("input.txt").expect("Unable to read file");
    let games = parse_game_data(&data);
    let mut sum = 0;
    for (game_index, game_data) in games {
        let result = test_game(&game_data);
        if result {
            sum += game_index
        }
    }
    println!("Sum: {}", sum);
}

fn get_power_of_game(game: &Vec<HashMap<String, i32>>) -> i32 {
    let mut current_cubes: HashMap<String, i32> = HashMap::new();
    current_cubes.insert("red".to_string(), 0);
    current_cubes.insert("green".to_string(), 0);
    current_cubes.insert("blue".to_string(), 0);
    for draw in game {
        for (color, count) in draw {
            let current_count = current_cubes.get_mut(color).unwrap();
            if (*current_count) < (*count) {
                *current_count = *count
            }
        }
    }
    return current_cubes.get("red").unwrap() * current_cubes.get("green").unwrap() * current_cubes.get("blue").unwrap()
}
fn solution_part_2() {
    let data = fs::read_to_string("input.txt").expect("Unable to read file");
    let games = parse_game_data(&data);
    let mut sum = 0;
    for (_, game_data) in games {
        sum += get_power_of_game(&game_data);
    }
    println!("Sum: {}", sum);
}
fn main() {
    solution_part_1();
    solution_part_2();
}
